[12.x] Add option to get typed command input

shaedrich opened the pull request:
and input validation

like, for example, with `\Illuminate\Support\Facades\Config` (#50140) and `\Illuminate\Support\Arr` (#55567)

# Enhancements
* return value is typed
* optionally provide a default value, otherwise, an exception is thrown if not found
* when value has unexpected type, an exception is thrown

# Example
```php
<?php

namespace App\Console\Commands;

use App\Models\User;
use App\Support\DripEmailer;
use Illuminate\Console\Command;
 
class SendEmails extends Command
{
    protected $signature = 'mail:send {user} {--flag} {--subject=} {--cc=*}';
    protected $description = 'Send a marketing email to a user';
 
    /**
     * Execute the console command.
     */
    public function handle(DripEmailer $drip): void
    {
        $user = User::find($this->integer('user')); // argument typed as int, alternatively, you can use $this->int() or $this->number()
        $flag = $this->boolean('flag'); // flag typed as bool, there's also the alias $this->bool()
        $subject = $this->string('subject', 'Default subject'); // option typed as string with default value
        $cc = User::whereIn('id', $this->array('cc'))->get(); // option typed as array

        if ($flag) {
            // do something
        }

        $drip->send($user, $cc, $subject);
    }
}
```

shaedrich committed:
Add option to get typed command input

and input validation

like, for example, with `\Illuminate\Support\Facades\Config`

nguyentranchung commented on src/Illuminate/Console/Concerns/InteractsWithIO.php:
```diff
@@ -460,4 +461,165 @@ public function outputComponents()
+     */
+    public function string(string $key, ?string $default = null): string
+    {
+        $value = match (true) {
```
This block is called multiple times.

shaedrich replied:
Ah, you mean, I could move that to its own method? Sure, I could, but I intend to first see if Taylor has an appetite for that

shaedrich committed:
StyleCI

shaedrich committed:
Remove `collection()`

as this already exists on `Illuminate\Foundation\Console\ResourceMakeCommand` for another purpose
and would cause a breaking change

shaedrich committed:
Remove unused import

Follow-up to 2067b7f

shaedrich committed:
StyleCI

shaedrich committed:
Add tests

shaedrich committed:
Use proper methods

shaedrich committed:
Test exceptions as well

shaedrich committed:
Add tests for other types

shaedrich committed:
Fix exception tests

shaedrich committed:
Fix test

shaedrich committed:
Add missing imports

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If applicable, please consider releasing your code as a package so that the community can still take advantage of your contributions!
